promql/engine.go

```diff
@@ -2811,19 +2820,37 @@ func vectorElemBinop(op parser.ItemType, lhs, rhs float64, hlhs, hrhs *histogram
 			}
 			return 0, res.Compact(0), true, nil
 		}
-		return lhs - rhs, nil, true, nil
+		if hlhs == nil && hrhs == nil {
+			return lhs - rhs, nil, true, nil
+		}
+		lhsTyp := "float"
+		rhsTyp := "histogram"
+		if hlhs != nil {
+			lhsTyp, rhsTyp = rhsTyp, lhsTyp
+		}
+		return 0, nil, false, annotations.NewIncompatibleTypesInBinOpInfo(lhsTyp, "-", rhsTyp, pos)
 	case parser.MUL:
 		if hlhs != nil && hrhs == nil {
 			return 0, hlhs.Copy().Mul(rhs), true, nil
 		}
 		if hlhs == nil && hrhs != nil {
 			return 0, hrhs.Copy().Mul(lhs), true, nil
 		}
+		if hlhs != nil && hrhs != nil {
+			return 0, nil, false, annotations.NewIncompatibleTypesInBinOpInfo("histogram", "*", "histogram", pos)
+		}
 		return lhs * rhs, nil, true, nil
 	case parser.DIV:
 		if hlhs != nil && hrhs == nil {
 			return 0, hlhs.Copy().Div(rhs), true, nil
 		}
+		if hrhs != nil {
+			lhsTyp := "float"
+			if hlhs != nil {
+				lhsTyp = "histogram"
+			}
+			return 0, nil, false, annotations.NewIncompatibleTypesInBinOpInfo(lhsTyp, "/", "histogram", pos)
```

Comment:
That's easier to read than the more complex cases above, but should we go with the explicit approach above, maybe we should do the same here for consistency. Your call.